functional_programming_tests/test/without_functional_programming.cpp

```diff
@@ -0,0 +1,205 @@
+	std::thread executor_thread_;
+
+	Map<unsigned char> map_;
+
+	rclcpp::Subscription<std_msgs::msg::UInt8MultiArray>::SharedPtr costmap_subscriber_;
+};
+
+// From the Robot, goal and costmap, generate a trajectory (Deterministic calculation)
+std::optional<Path> PathGenerator::generate_global_path(Pos const& start, Pos const& goal) { // Calculation
+    // Some cool and nifty algorithm
+    // What is the delta in position
+	int const del_x = goal.x - start.x;
+	int const del_y = goal.y - start.y;
+
+    // What direction to move in for each dimension
+	int const del_x_sign = std::copysign(1.0,del_x);
+	int const del_y_sign = std::copysign(1.0,del_y);
+
+    // Push start onto the path
+	Path path;
+	path.push_back(start);
+
+    // Fails if there is any obstacle in the way
+    // Move horizontally
+	for (size_t i = 0; i < (std::abs(del_x)); ++i) {
+		if (map_.at(path.back().y).at(path.back().x + del_x_sign) == 1) {
+			return std::nullopt;
+		}        
+		path.push_back({path.back().x+del_x_sign, path.back().y});
+	}
+    // Move vertically
+	for (size_t i = 0; i < (std::abs(del_y)); i++) {
+		if (map_.at(path.back().y + del_y_sign).at(path.back().x) == 1) {
+			return std::nullopt;
+		}            
+		path.push_back({path.back().x, path.back().y+del_y_sign});
+	}
+
+	return path;
+}
+
+// Function that gets costmap
+void PathGenerator::costmap_callback(const std_msgs::msg::UInt8MultiArray::SharedPtr msg) { // Action
+    // Get the costmap from a ros topic
+	// Check that map layout makes sense
+	if ((msg->layout.dim[0].size * msg->layout.dim[1].size) != msg->layout.dim[0].stride) {
+		RCLCPP_ERROR_STREAM(node_->get_logger(), "COSTMAP DIMENSIONS AND STRIDE INCONSISTENT!!");
+		return;
+	}
+	if (msg->layout.dim[0].stride != msg->data.size()) {
+		RCLCPP_ERROR_STREAM(node_->get_logger(), "COSTMAP LENGTH AND STRIDE INCONSISTENT!!");
+		return;
+	}
+	auto const begin = std::begin(msg->data) ;
+	// Populate the map
+	for (auto row = 0; row < msg->layout.dim[0].size; row++) {
+		map_.push_back({begin+row*msg->layout.dim[1].size, begin+(row+1)*msg->layout.dim[1].size});
+	}
+}
+
+class TaskPlanningFixture : public testing::Test {
+public:
+	// Adapted from minimal_integration_test
+	TaskPlanningFixture() : node_(std::make_shared<rclcpp::Node>("bad_test_publisher")), executor_(std::make_shared<rclcpp::executors::SingleThreadedExecutor>()) {
+        // Create ROS2 publisher for the costmap
+		costmap_publisher_ = node_->create_publisher<std_msgs::msg::UInt8MultiArray>("test/costmap",1);
+
+        // Publish the costmap every 100 ms
+		timer_ = node_->create_wall_timer(100ms, std::bind(&TaskPlanningFixture::publish_costmap, this));
+	}
+
+	void SetUp() override {
+		executor_->add_node(node_);
+		executor_thread_ = std::thread([this]() { executor_->spin(); });    
+	}
+
+    // Cleanup actions that could throw an exception
+	void TearDown() override {
+		executor_->cancel();
+		executor_thread_.join();
+	}
+
+protected:
```

Comment:
```suggestion
```
I haven't found a need to make things `protected` and it's generally a smell.

Reply:
Done